clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -2129,63 +2167,183 @@ private Map<String, KafkaFuture<TopicDescription>> handleDescribeTopicsByNames(f
+                                        partiallyFinishedTopicDescription = currentTopicDescription;
+                                    } else {
+                                        partiallyFinishedTopicDescription.partitions().addAll(currentTopicDescription.partitions());
+                                    }
+
+                                    if (!cursorTopicName.equals(topicName)) {
+                                        pendingTopics.remove(topicName);
+                                        future.complete(partiallyFinishedTopicDescription);
+                                        partiallyFinishedTopicDescription = null;
+                                    }
+                                    continue;
+                                }
+
+                                if (cursorTopicName.equals(topicName)) {
+                                    partiallyFinishedTopicDescription = currentTopicDescription;
+                                    continue;
+                                }
+
+                                pendingTopics.remove(topicName);
+                                future.complete(currentTopicDescription);
+                            }
+                            partiallyFinishedTopicName = cursorTopicName;
+                            partiallyFinishedTopicNextPartitionId = cursorPartitionId;
+                            if (pendingTopics.isEmpty()) {
+                                handleNonExistingTopics();
+                                nextRun.complete(false);
+                            } else {
+                                nextRun.complete(true);
+                            }
+                        }
+
+                        @Override
+                        boolean handleUnsupportedVersionException(UnsupportedVersionException exception) {
+                            return false;
+                        }
+
+                        @Override
+                        void handleFailure(Throwable throwable) {
+                            completeAllExceptionally(topicFutures.values(), throwable);
+                            nextRun.completeExceptionally(throwable);
+                        }
+                    };
+                }
+
+                void handleNonExistingTopics() {
+                    for (Map.Entry<String, KafkaFutureImpl<TopicDescription>> entry : topicFutures.entrySet()) {
+                        if (!entry.getValue().isDone()) {
+                            entry.getValue().completeExceptionally(new UnknownTopicOrPartitionException("Topic " + entry.getKey() + " not found."));
+                        }
                     }
-                    if (!cluster.topics().contains(topicName)) {
-                        future.completeExceptionally(new UnknownTopicOrPartitionException("Topic " + topicName + " not found."));
-                        continue;
+                }
+            };
+            call.run();
```

Comment:
It's my understanding that the reason we've got into this pagination business is because data from all partitions may not fit in memory at the same time, so we'd take streaming approach and retrieve data in chunks and then print it as arrives, so that we use O(1) memory instead of O(N).  Looking at the end-to-end workflow it looks like we still wait for all results to be collected in memory, so it doesn't look like we fixed the problem.

Reply:
I agree that option 3 would provide the most intuitive interface, I think 1 & 2 would hide streaming semantics behind an interface that is generally expected to be "in memory".